### PR TITLE
Add OverloadResolutionPriorityAttribute

### LINKED ISSUE #102173: [API Proposal]: `OverloadResolutionPriorityAttribute`
### Background and motivation

In order to support the C# proposed feature https://github.com/dotnet/csharplang/blob/d5bab90e846f9b3e9a7a4e552249643148f0194d/proposals/overload-resolution-priority.md, we add a new attribute to the BCL: `OverloadResolutionPriorityAttribute`. This feature allows applying the attribute to adjust the priority of an overload within a type. This will allow consumers like the rest of the BCL to adjust overload resolution for scenarios where a new, better API should be preferred (but won't be because overload resolution would prefer the existing one), or to resolve scenarios that will otherwise be ambiguous. One prime example is `Debug.Assert`; with this API, we could adjust `void Debug.Assert(bool condition, string? message)` to instead be `[OverloadResolutionPriority(1)] Debug.Assert(bool condition, [CallerArgumentExpression(nameof(condition))] string? message = "")`. The real change here would be a bit more complex due to the number of overloads that `Debug.Assert` already has, but this is a simple enough example to convey they intent.

### API Proposal

```csharp
namespace System.Runtime.CompilerServices;

[AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
public sealed class OverloadResolutionPriorityAttribute(int priority)
{
    public int Priority => priority;
}
```

### API Usage

```csharp
using System.Runtime.CompilerServices;

var d = new C1();
int[] arr = [1, 2, 3];
d.M(arr); // Prints "Span"

class C1
{
    [OverloadResolutionPriority(1)]
    public void M(ReadOnlySpan<int> s) => Console.WriteLine("Span");
    // Default overload resolution priority
    public void M(int[] a) => Console.WriteLine("Array");
}
```

### Alternative Designs

_No response_

### Risks

One of the BCL's main uses of `OverloadResolutionPriorityAttribute` will be to effectively make source-breaking changes while maintaining binary compatibility. Care will have to be taken when using the attribute to ensure only the intended consequences occur. For a few of these cases, it can effectively make specific overloads impossible to call directly, and users would have to resort to using something like delegate conversion, `UnsafeAccessor`, reflection, or manual IL.

### PR DESCRIPTION
**Pull request by 333fred**
Closes https://github.com/dotnet/runtime/issues/102173.
Relates to test plan https://github.com/dotnet/roslyn/issues/74131

### CODE REVIEW

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -13091,6 +13091,13 @@ public sealed partial class NullablePublicOnlyAttribute : System.Attribute
         public readonly bool IncludesInternals;
         public NullablePublicOnlyAttribute(bool value) { }
     }
+    [System.AttributeUsageAttribute(System.AttributeTargets.Method | System.AttributeTargets.Constructor | System.AttributeTargets.Property, Inherited=false)]
```
Comment (333fred) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:13094]:
```suggestion
    [System.AttributeUsageAttribute(System.AttributeTargets.Method | System.AttributeTargets.Constructor | System.AttributeTargets.Property, AllowMultiple=false, Inherited=false)]
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Specifies the priority of a member in overload resolution. When unspecified, the default priority is 0.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
+    public sealed class OverloadResolutionPriorityAttribute(int priority)
```
Comment (333fred) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs:10]:
```suggestion
    public sealed class OverloadResolutionPriorityAttribute(int priority) : Attribute
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Specifies the priority of a member in overload resolution. When unspecified, the default priority is 0.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
+    public sealed class OverloadResolutionPriorityAttribute(int priority) : Attribute
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs:10]: Nit: let's not use a primary constructor here and then put an XML comment on the constructor to document it.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        /// <summary>
+        /// The priority of the member.
+        /// </summary>
+        public int Priority { get; } = priority;
+    }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs:16]: Are there any constraints on what priority can be, or the entire range of Int32 is valid?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs:10]: As part of that it'd be good to clarify whether higher or lower is increased priority.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs (hunk 4 above)
Comment (333fred) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/OverloadResolutionPriorityAttribute.cs:16]: No constraints. Any Int32 is a valid value.

**Review by stephentoub — APPROVED** [review]
Thanks

**Pull request merged**